File: Sources/EwECustomPlugins/EwEValueChain/Data/Flow elements/Factories/cUnitFactory.vb
' SPDX-License-Identifier: EUPL-1.2
' This file is part of Ecopath with Ecosim (EwE).
' Copyright © 1991– Ecopath International Initiative (EII)

Public Class cUnitFactory

    Public Enum eUnitType As Integer
        All = 0
        Producer
        Processing
        Distribution
        Wholesaler
        Retailer
        Consumer
    End Enum

    Public Shared Function CreateUnit(tClass As Type) As cUnit
        If Not GetType(cUnit).IsAssignableFrom(tClass) Then Return Nothing
        Return CType(System.Activator.CreateInstance(tClass), cUnit)
    End Function

    Public Shared Function CreateUnit(unitType As eUnitType) As cUnit
        Return CreateUnit(MapType(unitType))
    End Function

    Public Shared Function MapType(unitType As eUnitType) As Type
        Dim t As Type = Nothing
        Select Case unitType
            Case eUnitType.Producer : t = GetType(cProducerUnit)
            Case eUnitType.Processing : t = GetType(cProcessingUnit)
            Case eUnitType.Distribution : t = GetType(cDistributionUnit)
            Case eUnitType.Wholesaler : t = GetType(cWholesalerUnit)
            Case eUnitType.Retailer : t = GetType(cRetailerUnit)
            Case eUnitType.Consumer : t = GetType(cConsumerUnit)
        End Select
        Return t
    End Function

    Public Shared Function CreateUnitDefault(unitType As eUnitType) As cUnit
        Dim t As Type = Nothing
        Select Case unitType
            Case eUnitType.Producer : t = GetType(cProducerUnitDefault)
            Case eUnitType.Processing : t = GetType(cProcessingUnitDefault)
            Case eUnitType.Distribution : t = GetType(cDistributionUnitDefault)
            Case eUnitType.Wholesaler : t = GetType(cWholesalerUnitDefault)
            Case eUnitType.Retailer : t = GetType(cRetailerUnitDefault)
            Case eUnitType.Consumer : t = GetType(cConsumerUnitDefault)
        End Select
        Return CType(System.Activator.CreateInstance(t), cUnit)
    End Function

End Class

File: Sources Internal/ValueChain/ValueChain/Data/Flow elements/Factories/cUnitFactory.cs
// ===============================================================================
// This file is part of Ecopath with Ecosim (EwE)
//
// EwE is free software: you can redistribute it and/or modify it under the terms
// of the GNU General Public License version 2 as published by the Free Software 
// Foundation.
//
// EwE is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; 
// without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR 
// PURPOSE. See the GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License along with EwE.
// If not, see <http://www.gnu.org/licenses/gpl-2.0.html>. 
//
//
// Copyright 1991- 
//    Ecopath International Initiative, Barcelona, Spain
// ===============================================================================

#region  Imports 

using System;

namespace ValueChain
{

    #endregion

    public class cUnitFactory
    {

        public enum eUnitType : int
        {
            All = 0,
            Producer,
            Processing,
            Distribution,
            Wholesaler,
            Retailer,
            Consumer
        }

        public static cUnit CreateUnit(Type tClass)
        {
            if (!typeof(cUnit).IsAssignableFrom(tClass))
                return null;
            return (cUnit)Activator.CreateInstance(tClass);
        }

        public static cUnit CreateUnit(eUnitType unitType)
        {
            return CreateUnit(MapType(unitType));
        }

        public static Type MapType(eUnitType unitType)
        {
            Type t = null;
            switch (unitType)
            {
                case eUnitType.Producer:
                    {
                        t = typeof(cProducerUnit);
                        break;
                    }
                case eUnitType.Processing:
                    {
                        t = typeof(cProcessingUnit);
                        break;
                    }
                case eUnitType.Distribution:
                    {
                        t = typeof(cDistributionUnit);
                        break;
                    }
                case eUnitType.Wholesaler:
                    {
                        t = typeof(cWholesalerUnit);
                        break;
                    }
                case eUnitType.Retailer:
                    {
                        t = typeof(cRetailerUnit);
                        break;
                    }
                case eUnitType.Consumer:
                    {
                        t = typeof(cConsumerUnit);
                        break;
                    }
            }
            return t;
        }

        public static cUnit CreateUnitDefault(eUnitType unitType)
        {
            Type t = null;
            switch (unitType)
            {
                case eUnitType.Producer:
                    {
                        t = typeof(cProducerUnitDefault);
                        break;
                    }
                case eUnitType.Processing:
                    {
                        t = typeof(cProcessingUnitDefault);
                        break;
                    }
                case eUnitType.Distribution:
                    {
                        t = typeof(cDistributionUnitDefault);
                        break;
                    }
                case eUnitType.Wholesaler:
                    {
                        t = typeof(cWholesalerUnitDefault);
                        break;
                    }
                case eUnitType.Retailer:
                    {
                        t = typeof(cRetailerUnitDefault);
                        break;
                    }
                case eUnitType.Consumer:
                    {
                        t = typeof(cConsumerUnitDefault);
                        break;
                    }
            }
            return (cUnit)Activator.CreateInstance(t);
        }

    }
}